File: solutions/4285-smallest-stable-index-ii/solution.py
class Solution:
    def firstStableIndex(self, nums: list[int], k: int) -> int:
        n = len(nums)
        minS = [float('inf')]*n
        minS[n-1] = nums[n-1]
        # print(minS)
        for i in range(n-2, -1, -1):
            # print(i, nums[i], minS[i])
            if nums[i] < minS[i+1]:
                minS[i] = nums[i]
            else:
                minS[i] = minS[i+1]
                
        
        maxS = -1
        for i in range(0, n):
            if nums[i] > maxS:
                maxS = nums[i]
            if (maxS-minS[i]) <= k:
                return i

        return -1 

File: solutions/4285-smallest-stable-index-ii/solution.java
class Solution {
    public int firstStableIndex(int[] nums, int k) {
        int n = nums.length;
        int[] minSuff = new int[n];
        minSuff[n-1] = nums[n-1];
        for(int i=n-2; i>=0; i--) {
            // System.out.println(i+ ": " + nums[i]);
            minSuff[i] = Math.min(nums[i], minSuff[i+1]);
        }
        int max = Integer.MIN_VALUE;
        for(int i=0; i<n; i++) {
            max = Math.max(max, nums[i]);
            // System.out.println(max + ", " + minSuff[i]);
            if(max - minSuff[i] <= k) {
                return i;
            }
        }
        return -1;
    }
}
